be/queue.c

```diff
@@ -0,0 +1,443 @@
+	bqop_tlist_move_tail(&bq->bq_op_get, bwo);
+	M0_LEAVE("bq="BEQ_F, BEQ_P(bq));
+}
+
+static void be_queue_op_get_done(struct m0_be_queue *bq, bool success)
+{
+	struct be_queue_wait_op *bwo;
+
+	M0_ENTRY("bq="BEQ_F" success=%d", BEQ_P(bq), !!success);
+	M0_PRE(m0_mutex_is_locked(&bq->bq_lock));
+	M0_PRE(!bqop_tlist_is_empty(&bq->bq_op_get));
+
+	bwo = bqop_tlist_head(&bq->bq_op_get);
+	if (success) {
+		be_queue_q_get(bq, &bwo->bbo_data, bwo->bbo_successful);
+	} else {
+		*bwo->bbo_successful = false;
+	}
+	m0_be_op_done(bwo->bbo_op);
+	bqop_tlist_move(&bq->bq_op_get_unused, bwo);
+	M0_LEAVE("bq="BEQ_F, BEQ_P(bq));
+}
+
+static bool be_queue_op_get_is_waiting(struct m0_be_queue *bq)
+{
+	return !bqop_tlist_is_empty(&bq->bq_op_get);
+}
+
+M0_INTERNAL void m0_be_queue_put(struct m0_be_queue  *bq,
+                                 struct m0_be_op     *op,
+                                 const struct m0_buf *data)
+{
+	struct be_queue_item *bqi;
+	bool                  was_full;
+
+	M0_ENTRY("bq="BEQ_F, BEQ_P(bq));
+	M0_PRE(m0_mutex_is_locked(&bq->bq_lock));
+	M0_PRE(be_queue_invariant(bq));
+	M0_PRE(!bq->bq_the_end);
+	M0_PRE(data->b_nob == bq->bq_cfg.bqc_item_length);
+
+	m0_be_op_active(op);
+	was_full = be_queue_is_full(bq);
+	bqi = be_queue_q_put(bq, data);
+	if (was_full) {
+		be_queue_op_put(bq, op, bqi);
+	} else {
+		m0_be_op_done(op);
+	}
+	/*
+	 * Shortcut for this case hasn't been not done intentionally.
+	 * It's much easier to look at the logs when all items are always added
+	 * to the queue.
+	 */
+	if (be_queue_op_get_is_waiting(bq))
+		be_queue_op_get_done(bq, true);
+
+	M0_POST(be_queue_invariant(bq));
+}
+
+M0_INTERNAL void m0_be_queue_end(struct m0_be_queue *bq)
+{
+	M0_ENTRY("bq="BEQ_F, BEQ_P(bq));
+	M0_PRE(m0_mutex_is_locked(&bq->bq_lock));
+	M0_PRE(be_queue_invariant(bq));
+	M0_PRE(!bq->bq_the_end);
+	M0_PRE(ergo(be_queue_op_get_is_waiting(bq), be_queue_is_empty(bq)));
+
+	bq->bq_the_end = true;
+	while (be_queue_op_get_is_waiting(bq))
+		be_queue_op_get_done(bq, false);
+	M0_POST(be_queue_invariant(bq));
+	M0_LEAVE("bq="BEQ_F, BEQ_P(bq));
+}
+
+M0_INTERNAL void m0_be_queue_get(struct m0_be_queue *bq,
+                                 struct m0_be_op    *op,
+                                 struct m0_buf      *data,
+                                 bool               *successful)
+{
+	M0_PRE(m0_mutex_is_locked(&bq->bq_lock));
+	M0_PRE(be_queue_invariant(bq));
+	M0_PRE(data->b_nob == bq->bq_cfg.bqc_item_length);
+
+	M0_ENTRY("bq="BEQ_F, BEQ_P(bq));
+	m0_be_op_active(op);
+	if (be_queue_is_empty(bq) && bq->bq_the_end) {
+		*successful = false;
+		m0_be_op_done(op);
+		return;
+	}
+	if (be_queue_is_empty(bq) || be_queue_op_get_is_waiting(bq)) {
+		be_queue_op_get(bq, op, data, successful);
+		return;
+	}
+	be_queue_q_get(bq, data, successful);
+	m0_be_op_done(op);
+	if (be_queue_op_put_is_waiting(bq))
+		be_queue_op_put_done(bq);
+	M0_POST(be_queue_invariant(bq));
+	M0_LEAVE("bq="BEQ_F, BEQ_P(bq));
+}
```

Comment:
How successful is assigned correctly?  Is successful input or output param

Reply:
It's an output parameter because it's a pointer to boolean. There is no reason to make an input parameter this way.